Onboarding: Add Import Process

emilylaguna opened the pull request:
| 📘 Project: #500 |
|:---:|

Adds the importing from other apps designs and steps. 

This also includes a `.shortcut` file but its not used. Since creating a shortcut requires me to share it via iCloud to be installed, I wanted to be able to include the "source" and that's what this file allows. 

## Screenshots
| List | Apple Podcasts | Breaker | Castro | Castbox | Overcast | Other Apps |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|<img src="https://user-images.githubusercontent.com/793774/202339163-fd0b1054-0e3e-4138-8040-c225fd50b160.png" height="200" />|<img src="https://user-images.githubusercontent.com/793774/202338813-92fc4b41-04de-462d-8c5d-9af6e9c3996d.png" height="200" />|<img src="https://user-images.githubusercontent.com/793774/202338817-640c3f86-2fa9-4ab9-9f74-587dc9efc2c8.png" height="200" />|<img src="https://user-images.githubusercontent.com/793774/202338818-4ef62218-de63-497d-b612-245f3c5ca287.png" height="200" />|<img src="https://user-images.githubusercontent.com/793774/202338820-0cb575dc-a1fe-443e-a2e6-9b5b1e8ef407.png" height="200" />|<img src="https://user-images.githubusercontent.com/793774/202338822-5bcb4683-ec8e-4cb3-8e13-beb15d4b1b25.png" height="200" />|<img src="https://user-images.githubusercontent.com/793774/202338823-54a45eeb-d9a5-4017-b2de-9e68f0ef77c0.png" height="200" />|

<details>
  <summary><h2>Apple Podcasts import demo</h2></summary>
  
https://user-images.githubusercontent.com/793774/202340250-bcf49d18-0cca-4ef4-ab29-4fc0fe6858fc.MP4
</details>

## To test

> 📱 **You will need to test on a real device** 📱

### Design Check

1. Launch the app
2. Go to the Podcasts tab
3. Tap the Create Folder button
4. Tap the Import Podcasts button
5. ✅ Verify the import list [matches the designs](https://www.figma.com/file/PviUP0aiZctOprDuuYRwpb/Onboarding-%26-Upgrade-v2?node-id=407%3A26265&t=07COUyjdz9xdTbeh-4)
6. Tap on any item
7. ✅ Verify the detail view [matches the designs](https://www.figma.com/file/PviUP0aiZctOprDuuYRwpb/Onboarding-%26-Upgrade-v2?node-id=407%3A26061&t=07COUyjdz9xdTbeh-4)

### App Detection
1. Install the following apps from the App Store: Breaker, Castro, Castbox, and Overcast
2. Launch the app
3. Go to the Podcasts tab
4. Tap the Create Folder button
5. Tap the Import Podcasts button
6. ✅ Verify you see all of the supported apps listed
7. Dismiss the view
8. Delete one of the apps
9. Go back through the flow to open the list
10. ✅ Verify the app is no longer listed

### No Apps Installed
5. Delete all the supported apps including Apple Podcasts
6. OR open the ImportViewModel and return false on line 33
1. Launch the app
2. Go to the Podcasts tab
3. Tap the Create Folder button
4. Tap the Import Podcasts button
7. ✅ Verify you see the "importing from other apps" steps directly and not the list

### Importing from Apple Podcasts
1. Launch the app
2. Go to the Podcasts tab
3. Tap the Create Folder button
4. Tap the Import Podcasts button
5. Tap on the Apple Shortcuts item
6. 📖 Read the steps and verify they make sense and do not have any typos
7. Tap the Install Shortcut Button
8. ✅ Verify you are brought to the Apple Shortcuts app and are presented with the import shortcut to install
9. Tap the Add Shortcut button
10. Tap the shortcut in the list to run it
11. ✅ Verify after its finished you are brought back to the app and the podcasts you had in Apple Podcasts are imported correctly

### Importing from supported apps
1. Launch the app
2. Go to the Podcasts tab
3. Tap the Create Folder button
4. Tap the Import Podcasts button
5. Go through each app, 📖 Read the steps and verify they make sense and do not have any typos
6. Follow the steps and import an OPML into the apps
7. ✅ Verify when the file is imported the import flow is no longer visible

## Checklist

- [ ] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

emilylaguna committed:
Add the Apple Podcasts to Pocket Casts shortcut file

emilylaguna committed:
Add image assets

emilylaguna committed:
Add more types of OPML that we support

emilylaguna committed:
Add a way to unsubscribe from all podcasts in the developer menu

emilylaguna committed:
Add localized strings

emilylaguna committed:
Fix an issue with the motion still being used even when the view is hidden

emilylaguna committed:
Rename to viewModel

emilylaguna committed:
Push to the import view from the welcome view

emilylaguna committed:
Add import hosting vc

emilylaguna committed:
Add import view model

emilylaguna committed:
Add import details view

emilylaguna committed:
Add import landing view

emilylaguna committed:
Update the test welcome view showing

emilylaguna committed:
Update the scroll view if needed to allow it to be recalculated

emilylaguna committed:
Add files

emilylaguna committed:
Remove debug item

emilylaguna committed:
Add import apps to queried urls

emilylaguna committed:
Fix some small device and ios 14 issues

emilylaguna committed:
Fix some issues with opening apps

emilylaguna committed:
Add castbox to the plist

emilylaguna commented on podcasts/AppDelegate+UrlHandling.swift:
```diff
@@ -27,7 +27,12 @@ extension AppDelegate {
             guard let type = uti?.takeRetainedValue() else { return false }
 
-            if UTTypeConformsTo(type, kUTTypeXML) || UTTypeConformsTo(type, "unofficial.opml" as CFString) {
+            let supportedTypes = [kUTTypeXML, "unofficial.opml" as CFString, "public.opml" as CFString]
```
The more common OPML UTI is public.opml so this adds support for it

emilylaguna commented on podcasts/DeveloperMenu.swift:
```diff
@@ -12,6 +13,14 @@ struct DeveloperMenu: View {
                     NotificationCenter.postOnMainThread(notification: Constants.Notifications.chartRegionChanged)
                 }
+
+                Button("Unsubscribe from all Podcasts") {
```
Needed this for my own testing.

emilylaguna commented on podcasts/Onboarding/Import/ImportDetailsView.swift:
```diff
@@ -0,0 +1,53 @@
+        let lines = app.steps.split(separator: "\n").map { String($0).trim() }
+        VStack(alignment: .leading, spacing: 20) {
+           ForEach(lines, id: \.self) { line in
+               Text(line)
```
I liked the idea of having a step by step list for Voice over users instead of a giant wall of text. So I am generating the list manually here.

emilylaguna committed:
Remove duplicate code

emilylaguna committed:
Make buttons clicky

emilylaguna commented on podcasts/Onboarding/Import/ImportViewModel.swift:
```diff
@@ -0,0 +1,91 @@
+    enum ImportAppId: String {
+        case breaker, castbox = "wazecastbox", overcast, other
+        case castro = "co.supertop.Castro-2"
+        case applePodcasts = "https://www.icloud.com/shortcuts/d9e0793e40ed4b5d9dd78c81e6af9234"
```
We have to share the URL here.

emilylaguna commented on podcasts/Onboarding/Login/LoginLandingView.swift:
```diff
@@ -143,7 +143,12 @@ private struct LoginHeader: View {
                                   topPadding: topPadding,
                                   manager: motion)
             }
+        }.onDisappear() {
```
Oops, fixes an issue where the motion never turned off.

emilylaguna committed:
Change back to plus

emilylaguna commented on podcasts/PodcastListViewController.swift:
```diff
@@ -220,12 +220,22 @@ class PodcastListViewController: PCViewController, UIGestureRecognizerDelegate,
 
     @objc private func createFolderTapped(_ sender: UIBarButtonItem) {
         if FeatureFlag.onboardingUpdates {
```
Temporary logic, will be removed.

emilylaguna commented on podcasts/podcasts-Info.plist:
```diff
@@ -382,10 +382,10 @@
-                <array>
-                    <string>AppIcon-Halloween</string>
-                </array>
+				<key>CFBundleIconFiles</key>
```
Thanks Xcode for the random changes.

pocketcasts commented:
<!--
  0 Errors
  0 Warnings
  0 Messages
  1 Markdown
-->

### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
ScrollViewIfNeeded.swift | 13 | Lines should not have trailing whitespace. (trailing_whitespace)


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

emilylaguna committed:
Change shadows

leandroalonso reviewed (approved):
Nice work on this!

I have two non-blockers issues though:

* The whole row is not touchable (when listing the apps) and the same goes for "Import Podcasts"
* I always see "Import from Apple Podcasts" even after I deleted it

emilylaguna committed:
Revert scrollview if needed

emilylaguna committed:
Make the entire button tappable

emilylaguna commented:
> The whole row is not touchable (when listing the apps) and the same goes for "Import Podcasts"

Ah forgot the content shape. I fixed that here: 10a38b9b975d80177a9c311926d62c0b824d3f2e

> I always see "Import from Apple Podcasts" even after I deleted it

Hmm yeah looks like the Apple apps are treated differently and "always" installed regardless. Not too worried about it.
